finality-aleph/src/new_network/connection_manager/session.rs

```diff
@@ -0,0 +1,417 @@
+use crate::{
+    crypto::KeyBox,
+    new_network::{
+        connection_manager::{
+            get_unique_peer_id, is_global, is_p2p, AuthData, Authentication, Multiaddr,
+        },
+        PeerId,
+    },
+    NodeIndex, SessionId,
+};
+use aleph_bft::{Index, KeyBox as _, NodeCount};
+use codec::Encode;
+use std::collections::HashMap;
+
+/// A struct for handling authentications for a given session and maintaining
+/// mappings between PeerIds and NodeIndexes within that session.
+pub struct Handler {
+    peers_by_node: HashMap<NodeIndex, PeerId>,
+    authentications: HashMap<PeerId, (Authentication, Option<Authentication>)>,
+    own_authentication: Authentication,
+    own_peer_id: PeerId,
+    keychain: KeyBox,
+}
+
+/// Returned when a set of addresses is not usable for creating authentications.
+/// Either because none of the addresses are externally reachable libp2p addresses,
+/// or the addresses contain multiple libp2p PeerIds.
+#[derive(Debug)]
+pub enum AddressError {
+    NoExternalP2pAddresses,
+    MultiplePeerIds,
+}
+
+async fn construct_authentication(
+    keychain: &KeyBox,
+    session_id: SessionId,
+    addresses: Vec<Multiaddr>,
+) -> Result<(Authentication, PeerId), AddressError> {
+    let addresses: Vec<_> = addresses
+        .into_iter()
+        .filter(is_global)
+        .filter(is_p2p)
+        .collect();
+    if addresses.is_empty() {
+        return Err(AddressError::NoExternalP2pAddresses);
```

Comment:
In a conversation with @timorl we realized that there might be issues with local networks if we try to do authentications for external addresses only. Leaving this comment for the record only. This is being worked on by @timorl now.

Reply:
Removed the filtering on global addresses, it will instead happen on the `add_reserved` level, since there it shouldn't matter for local tests.